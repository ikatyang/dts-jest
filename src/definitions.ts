import * as _ts from 'typescript';

// tslint:disable-next-line:no-var-requires
export const package_name = require('../package.json').name;

export const config_namespace = '_dts_jest_';
export const runtime_namespace = '_dts_jest_runtime_';

export const env_root_dir = 'DTS_JEST_ROOT_DIR';

export const runtime_indent_spaces = 2;

export const trigger_regex = /^\s*\/\/\s*@dts-jest\b(:?\S*)\s*(.+)?\s*$/;
export enum TriggerMatchIndex {
  Input,
  Flags,
  Description,
}
export type TriggerMatchArray = [string, string, string | undefined];

export enum AssertionFlag {
  Snapshot = ':snapshot',
  Show = ':show',
  Pass = ':pass',
  Fail = ':fail',
}
export enum ActualAssertionFlag {
  Error = ':error',
  NoError = ':no-error',
  Question = '?',
}

export enum TestFlag {
  Test = ':test',
  Only = ':only',
  Skip = ':skip',
}
export enum TestMethod {
  Test = 'test',
  Only = 'test.only',
  Skip = 'test.skip',
}
export const group_flag = ':group';
export enum GroupMethod {
  Test = 'describe',
  Only = 'describe.only',
  Skip = 'describe.skip',
}

export interface Target {
  /**
   * 0-based
   */
  line: number;
  expression: string;
  description?: string;
  group?: Group;
}

export interface Snapshot {
  inference?: string;
  diagnostic?: string;
}
export interface Expected {
  value: string;
}

export interface TriggerInfo {
  flags: AssertionFlag[];
  method: TestMethod;
}
export interface Trigger extends Target, TriggerInfo {
  start: number;
  end: number;
}
export interface ActualTrigger extends Trigger, Expected {}

export interface GroupInfo {
  method: GroupMethod;
}
export interface Group extends GroupInfo {
  title: string;
}

export type TriggerOrGroupInfo =
  | ({ is_group: false } & TriggerInfo)
  | ({ is_group: true } & GroupInfo);

export interface Result extends Target, Snapshot {}
export interface ActualResult extends Target, Expected {}

export interface JestConfig {
  rootDir: string;
  globals: { [K in typeof config_namespace]?: RawConfig };
  _dts_jest_internal_test_?: boolean;
}

export interface RawConfig {
  tsconfig?: string | Record<string, _ts.CompilerOptionsValue>;
  type_format?: _ts.TypeFormatFlags;
  typescript?: string;
}

export interface FormattedConfig {
  tsconfig: _ts.CompilerOptions;
  type_format: _ts.TypeFormatFlags;
  typescript: typeof _ts;
  typescript_path: string;
}

export type JestSnapshot = Record<string, string>;
