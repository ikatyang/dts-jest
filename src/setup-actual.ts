import { ActualResult } from './definitions';
import { ActualRuntime } from './runtime-actual';

export const setup_actual = (actual_results: ActualResult[]) => {
  const results = actual_results.reduce<{ [line: number]: ActualResult }>(
    (current_results, result) => ({
      ...current_results,
      [result.line]: result,
    }),
    {},
  );

  return new ActualRuntime(results);
};
