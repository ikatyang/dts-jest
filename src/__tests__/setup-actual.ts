import { ActualResult } from '../definitions';
import { setup_actual } from '../setup-actual';

const results: ActualResult[] = [
  { line: 0, expression: 'first', value: '123' },
  { line: 3, expression: 'second', value: '456' },
];

it('should setup correctly', () => {
  expect(setup_actual(results)).toMatchSnapshot();
});
