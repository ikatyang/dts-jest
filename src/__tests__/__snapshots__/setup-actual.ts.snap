// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should setup correctly 1`] = `
ActualRuntime {
  "results": Object {
    "0": Object {
      "expression": "first",
      "line": 0,
      "value": "123",
    },
    "3": Object {
      "expression": "second",
      "line": 3,
      "value": "456",
    },
  },
}
`;
